src/sage/modular/hypergeometric_motive.py

```diff
@@ -1751,25 +1931,85 @@ def euler_factor(self, t, p, cache_p=False):
         t = QQ(t)
         if t in [0, 1]:
             raise ValueError('invalid t')
-        alpha = self._alpha
-        if 0 in alpha:
-            return self._swap.euler_factor(~t, p)
-
         if not is_prime(p):
             raise ValueError('p not prime')
         if not all(x.denominator() % p for x in self._alpha + self._beta):
             raise NotImplementedError('p is wild')
-        if (t.numerator() * t.denominator() % p == 0 or (t - 1) % p == 0):
-            raise NotImplementedError('p is tame')
-        # now p is good
-        d = self.degree()
+        if 0 in self._alpha:
+            return self._swap.euler_factor(~t, p)
+        P = PolynomialRing(ZZ, 'T')
+        if t.numerator() % p == 0 or t.denominator() % p == 0:
+            ans = P.one()
+            for m in set(j for i in self.cyclotomic_data() for j in i):
+                ans *= self.euler_factor_tame_contribution(t, p, m, deg)
+            if deg is not None:
+                ans = ans.truncate(deg+1)
+            return ans
+        # now p is good, or p is tame and t is a p-adic unit
+        elif (t-1) % p == 0:
+            typ = "mult"
+            d = (self.degree() - 1) // 2 * 2
```

Comment:
I still find it takes me a moment to realize that these operations should not simply cancel (and could follow my suggestion on the first round).

Reply:
Maybe it's better to be clear than concise, reworked the logic.